### PR TITLE
Fix/lsp/iterators completion

### PR DESCRIPTION
**Pull request by hugocaillard**
Fix: #865 

This PR aims to give more context when auto completing the first argument (callback) of the iterator function (map, filter, fold). It's relevant in two cases:
- when the auto completion is triggered for one of these 3 functions, the LSP provides a list of choices for the first argument
- when the developer comes back later to edit the first argument, the LSP used to provide a generic list of completion items, it now only gives the functions that could fit here

Some other fixes were added to this PR:
- user defined functions with no argument wouldn't be listed in completion items
- smart parenthesis wrap (`should_wrap`) would only work if the cursor was at the end of the line
- define-private signature help wasn't skipped

Part of the code/logic (especially hard coded stuff like list of functions) will be reorganised and move somewhere else in the context of #691 in Q2

⚠️ These changes won't be released without https://github.com/stacks-network/stacks-blockchain/pull/3569

### CODE REVIEW

**Review by obycode — DISMISSED** [review]
LGTM!

File: components/clarity-lsp/src/common/requests/completion.rs
Comment (hugocaillard) [reply components/clarity-lsp/src/common/requests/completion.rs:100]: Yes i initially tried to favorise empty vecs.

This method kind of does two things (the second being populating function_completion_items), and because of the `if let Some(parameters)` I had before, it wasn't called for functions with no arguments. So this was the simple way to handle it.
I might need to split it into two functions. I'll give it a try before merging

**Review by obycode — DISMISSED** [review]
lgtm!

**Comment by hugocaillard** [comment]
@obycode Just saw that my build was broken, I updated `lsp-types` but `towser-lsp` doesn't have the last version so yet, I have to keep both in sync. Need a new approval

**Review by obycode — APPROVED** [review]

**Pull request merged**
